wallet/key_deriver.go

```diff
@@ -0,0 +1,153 @@
+package wallet
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	ec "github.com/bsv-blockchain/go-sdk/primitives/ec"
+)
+
+type KeyDeriver struct {
+	privateKey *ec.PrivateKey
+	publicKey  *ec.PublicKey
+}
+
+func NewKeyDeriver(privateKey *ec.PrivateKey) *KeyDeriver {
+	return &KeyDeriver{
+		privateKey: privateKey,
+		publicKey:  privateKey.PubKey(),
+	}
+}
+
+func (kd *KeyDeriver) DeriveSymmetricKey(protocol WalletProtocol, keyID string, counterparty WalletCounterparty) ([]byte, error) {
+	// Prevent deriving symmetric key for self
+	if counterparty.Type == CounterpartyTypeSelf {
+		return nil, fmt.Errorf("cannot derive symmetric key for self")
+	}
+
+	// If counterparty is 'anyone', use a fixed public key
+	if counterparty.Type == CounterpartyTypeAnyone {
+		_, fixedKey := ec.PrivateKeyFromBytes([]byte{1})
+		counterparty = WalletCounterparty{
+			Type:         CounterpartyTypeOther,
+			Counterparty: fixedKey,
+		}
+	}
+
+	// Derive both public and private keys
+	derivedPublicKey, err := kd.DerivePublicKey(protocol, keyID, counterparty, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive public key: %w", err)
+	}
+
+	derivedPrivateKey, err := kd.DerivePrivateKey(protocol, keyID, counterparty)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive private key: %w", err)
+	}
+
+	// Create shared secret
+	sharedSecret, err := derivedPrivateKey.DeriveSharedSecret(derivedPublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create shared secret: %w", err)
+	}
+	if sharedSecret == nil {
+		return nil, fmt.Errorf("failed to derive shared secret")
+	}
+
+	// Return the x coordinate of the shared secret point
+	return sharedSecret.X.Bytes(), nil
+}
+
+func (kd *KeyDeriver) DerivePublicKey(protocol WalletProtocol, keyID string, counterparty WalletCounterparty, forSelf bool) (*ec.PublicKey, error) {
+	counterpartyKey := kd.NormalizeCounterparty(counterparty)
+	invoiceNumber, err := kd.ComputeInvoiceNumber(protocol, keyID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compute invoice number: %w", err)
+	}
+
+	var pubKey *ec.PublicKey
+	if forSelf {
+		privKey, err := kd.privateKey.DeriveChild(counterpartyKey, invoiceNumber)
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive child private key: %w", err)
+		}
+		pubKey = privKey.PubKey()
```

Comment:
Minor thing:
maybe this one, could be simply returned, and the else part could be unwrapped then.

Reply:
I agree, I like that pattern better.